### DISCUSSION TITLE
Issues with the mouse

### DISCUSSION THREAD

**Discussion by vruumllc** [picocomputer/community#79] [Q&A]
Hi, I'm trying to the mouse to control the cursor a simple text mode (mode1) editor that uses both the keyboard and the mouse to move the cursor. I'm using canvas mode3 (640X480), but I think the issues will still happen in 320x240.

Problem 1: If I move the cursor with mouse, then keyboard, then mouse, the cursor jumps back to original mouse position.  I need a way to set mouse position to the current cursor position set with keyboard.

I tried writing to the mouse RIA registers like this:
    RIA.addr0 = MOUSE_INPUT + 1;
    RIA.step0 = 0;
    RIA.rw0 = cur_col * 8;
    RIA.addr0 = MOUSE_INPUT + 2;
    RIA.step0 = 0;
    RIA.rw0 = cur_row * 16;

Problem 2: This doesn't work, however, because the x and y mouse position registers are only a byte, 255 max, and the current position on the screen can exceed that in both X and Y.  

In the paint program you used static int sx, sy to accumulate small delta movements, which is why it worked there, but I can't go backwards that way.

Any suggestions?

Thanks, tonyvr

**Comment by vruumllc** [comment]
Stupid web page...The first sentence should read:

Hi, I'm trying to write a simple text mode (mode1) editor that uses both the keyboard and the mouse to move a text mode cursor.

**Comment by rumbledethumps** [comment]
Writing to the mouse registers should be considered undefined behavior.

For even a trivial use case of the mouse, you'll have to implement a small driver that works like the paint program. Accumulate the small char movements and apply their delta to your cursor position. Perhaps keep your cursor position as a fixed point 8.8 so you can slow down the mouse as necessary.

This works like a Commodore 64 mouse. Maybe you can get some ideas looking at what C64 programmers do.

**Reply by vruumllc** [reply]
I've decided to abandon trying to use a single blinking text mode char as both a cursor and mouse pointer. This is what Norton Commander did back in the day, and I thought it might be simpler than using two graphics modes and having separate text cursor and graphical mouse.

I will just use a separate graphical pointer that moves the text mode cursor upon a left button click, like VSCode.  That is the behavior we are all used to these days anyway.

Thanks for the quick response. Limiting the mouse X, Y registers to only 8 bits still seems like it might present problems, though, since the screen resolutions exceed that.

**Reply by rumbledethumps** [reply]
The paint example has no problem with 320 pixels across. You're looking at those values as absolutes when they are meant for you to grab a delta from once in a while (for values of while >= 60Hz).

Think: "Ok, the x_mouse moved -2 since last I looked, so now I apply -2 * speed to my fractional x_position."

**Reply by vruumllc** [reply]
I just checked, and it turns out I misremembered how Norton Commander worked anyway. It used a text mode red block as the mouse, and blinking underscore attribute as the cursor, so there were indeed two separate positions to keep track of. You had to click the mouse to reset the text mode cursor position, just as we do today.

**Reply by vruumllc** [reply]
As for the way the paint mouse code works, I understand that.  (or at least I thought I did). It just seems like the x and y variables can roll over and you can get an incorrect delta  computed from your last saved value. I'll think about it some more.

**Reply by rumbledethumps** [reply]
The USB protocol is 8 bit and prone to the same rollover. The faster you poll, the faster the user can move the mouse. Most mice these days report 250Hz and higher, which a 6502 application polling >= 250Hz can take advantage of. There isn't a way to give the 6502 more data or faster reports.
In practice, 60Hz will rollover on mouse movements that were impossibly fast on "ball" mice. It's fine, but more is better.

**Reply by rumbledethumps** [reply]
Oh, and if you're wondering why the RIA doesn't track and report in 16 bits for slower polls, it's because there's no way to atomically transfer more than 8 bits. Unironically, the same problems we had to code around back in the day are still here.

The vsync example shows how to do interrupts on vsync, which you can write a mouse driver in. If it's not fast enough for you, use a 6522 timer to make a faster interrupt.

**Reply by vruumllc** [reply]
Interesting stuff. Thanks for your time.
I have been using your code from paint, and it does seem to work fine. I always like to think about the the edge cases, though. It's fun!
